File: iomerge-server-app/src/main/java/com/github/krzychek/server/ui/TrayManager.java
package com.github.krzychek.server.ui;

import org.pmw.tinylog.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import javax.imageio.ImageIO;
import java.awt.AWTException;
import java.awt.MenuItem;
import java.awt.PopupMenu;
import java.awt.SystemTray;
import java.awt.TrayIcon;


@Component
class TrayManager {

	private final SettingsWindow settingsWindow;
	private SystemTray tray;
	private TrayIcon trayIcon;

	@Autowired
	public TrayManager(SettingsWindow settingsWindow) {
		this.settingsWindow = settingsWindow;

		if (!SystemTray.isSupported()) {
			Logger.error("Tray is not supported on this system");
			return;
		}
		tray = SystemTray.getSystemTray();
		try {
			trayIcon = new TrayIcon(ImageIO.read(getClass().getResource("/icon.png")), "IOMerge");
			trayIcon.setImageAutoSize(true);
		} catch (Exception e) {
			Logger.error("filed to load tray icon", e);
			return;
		}
		createAndShow();
	}

	private void createAndShow() {
		tray.remove(trayIcon);

		PopupMenu popup = new PopupMenu();
		popup.add(new MenuItem("Settings")).addActionListener(e -> settingsWindow.show());
		popup.add(new MenuItem("Exit")).addActionListener(e -> Runtime.getRuntime().exit(0));

		trayIcon.setPopupMenu(popup);

		try {
			tray.add(trayIcon);
		} catch (AWTException e) {
			System.out.println("TrayIcon could not be added.");
		}
	}
}

File: iomerge-server-app/src/main/java/com/github/krzychek/iomerge/server/ui/TrayManager.kt
package com.github.krzychek.iomerge.server.ui

import com.github.krzychek.iomerge.server.daggerConfig.LifecycleManager
import dagger.Lazy
import dorkbox.systemTray.SystemTray
import org.pmw.tinylog.Logger
import java.awt.MenuItem
import java.awt.PopupMenu
import java.awt.TrayIcon
import javax.imageio.ImageIO
import javax.inject.Inject
import javax.inject.Singleton


@Singleton class TrayManager
@Inject constructor(lazySettingsWindow: Lazy<SettingsWindow>, val lifecycleManager: LifecycleManager) {

	val settingsWindow: SettingsWindow by lazy { lazySettingsWindow.get() }


	private fun tryLoadNativeTray() {
		SystemTray.getSystemTray()!!.apply {

			status = "IOMerge"

			setIcon("trayIcon", javaClass.getResourceAsStream("/icon.png"))

			addMenuEntry("Settings", { systemTray, menuEntry -> settingsWindow.show() })
			addMenuEntry("Exit", { systemTray, menuEntry -> lifecycleManager.shutdown() })
		}
	}

	private fun loadAwtTray() {

		val image = ImageIO.read(javaClass.getResource("/icon.png"))

		val trayIcon = TrayIcon(image, "IOMerge").apply {
			isImageAutoSize = true
			popupMenu = PopupMenu().apply {
				add(MenuItem("Settings")).addActionListener { settingsWindow.show() }
				add(MenuItem("Exit")).addActionListener { lifecycleManager.shutdown() }
			}
		}
		java.awt.SystemTray.getSystemTray().add(trayIcon)
	}

	init {

		var loaded = false
		try {
			tryLoadNativeTray()
			loaded = true
		} catch (e: Throwable) {
			Logger.warn(e, "Problems with loading native tray, trying awt.")
		}

		if (!loaded) try {
			loadAwtTray()
		} catch (e: Throwable) {
			Logger.warn(e, "Filed to load awt tray")
		}
	}
}